purchase_sale_inter_company/models/purchase_order.py

```diff
@@ -103,30 +103,25 @@ def _inter_company_create_sale_order(self, dest_company):
         if dest_company.sale_auto_validation:
             sale_order.with_user(intercompany_user.id).sudo().action_confirm()
 
-    def _prepare_sale_order_data(
-        self, name, partner, dest_company, direct_delivery_address
-    ):
+    def _prepare_sale_order_data(self, name, dest_company, direct_delivery_address):
         """Generate the Sale Order values from the PO
         :param name : the origin client reference
         :rtype name : string
-        :param partner : the partner reprenseting the company
-        :rtype partner : res.partner record
         :param dest_company : the company of the created SO
         :rtype dest_company : res.company record
         :param direct_delivery_address : the address of the SO
         :rtype direct_delivery_address : res.partner record
         """
+        # partner_id on new order should be partner from destanation company
+        # if not: "Incompatible companies on records" exception is raised
+        # during related invoice creation
         self.ensure_one()
-        delivery_address = (
-            direct_delivery_address
-            or self.picking_type_id.warehouse_id.partner_id
-            or False
-        )
+        delivery_address = direct_delivery_address or dest_company.partner_id or False
         new_order = self.env["sale.order"].new(
             {
                 "company_id": dest_company.id,
                 "client_order_ref": name,
-                "partner_id": partner.id,
+                "partner_id": dest_company.partner_id.id,
                 "date_order": self.date_approve,
                 "auto_purchase_order_id": self.id,
             }
```

Comment:
@gurneyalex can you pls take a look, I have doubts about it.
1. `delivery_address = direct_delivery_address or dest_company.partner_id` looks correct to me as we specify destination address on new generated SO
2. `partner_id` on new SO changed to destination company partner. On Invoice creation `Incompatible companies on records` is raised if partner is the same as on the PO. So, for example we got Invoice with company B and partner with company A.